packages/cubejs-api-gateway/src/gateway.ts

```diff
@@ -1040,26 +1040,73 @@ class ApiGateway {
     const requestStarted = new Date();
     try {
       if (!query) {
-        throw new UserError(
-          'A user\'s query must contain a body'
-        );
+        throw new UserError('A user\'s query must contain a body');
       }
 
       if (!Array.isArray(query) && !query.query) {
         throw new UserError(
           'A user\'s query must contain at least one query param.'
         );
       }
-
+      
       query = {
         ...query,
-        requestId: context.requestId
+        requestId: context.requestId,
       };
 
-      if (query.resultFilter?.objectTypes && !Array.isArray(query.resultFilter.objectTypes)) {
-        throw new UserError(
-          'A query.resultFilter.objectTypes must be an array of strings'
-        );
+      // Determine SQL query type and add LIMIT clause if needed
+      const shouldAddLimit = (sql: string): Boolean => {
+        // TODO: Enhance the way we determine query type
+        const ddlRegex = /^(CREATE|ALTER|DROP|TRUNCATE)\b/i;
+        const dmlRegex = /^(INSERT|UPDATE|DELETE)\b/i;
+        const selectRegex = /^(SELECT|WITH)\b/i;
+      
+        if ((ddlRegex.test(sql)) || (dmlRegex.test(sql))) {
+          return false;
+        } else if (selectRegex.test(sql)) {
+          return true;
+        } else {
+          throw new UserError(
+            'Invalid SQL query'
+          );
+        }
+      };
+
+      if (shouldAddLimit(query.query)) {
+        if (
+          !query.limit ||
+          !Number.isInteger(query.limit) ||
+          query.limit > 50000 ||
+          query.limit < 1
+        ) {
+          throw new UserError(
+            'A user\'s query must contain limit query param and it must be positive number less than 50000.'
```

Comment:
We've provided env variables to manage max query limit: `getEnv('dbQueryDefaultLimit')` and `getEnv('dbQueryLimit')`. Check `ApiGateway#processQueryLimit()` method for the details. Please replace hardcoded constant.